contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,193 @@
+        args: ["token", "-i", "(CLUSTER_ID)", "-r", "(ROLE_ARN)"]
+```
+
+## TLS client certificate support
+
+The current The auth provider interface doesn't let the user modify the dialer,
+only wrap the transport.
+
+```
+type AuthProvider interface {
+    // WrapTransport allows the plugin to create a modified RoundTripper that
+    // attaches authorization headers (or other info) to requests.
+    WrapTransport(http.RoundTripper) http.RoundTripper
+    // Login allows the plugin to initialize its configuration. It must not
+    // require direct user interaction.
+    Login() error
+}
+```
+
+Since this doesn't let a `AuthProvider` supply things like client certificates,
+the signature of the `AuthProvider` should change too ([with corresponding changes
+to `k8s.io/client-go-transport`](
+https://gist.github.com/ericchiang/7f5804403b359ebdf79dcf76c4071bff)):
+
+```
+import (
+    "k8s.io/client-go/transport"
+    // ...
+)
+
+type AuthProvider interface {
+    // UpdateTransportConfig updates a config by adding a transport wrapper,
+    // setting a bearer token (should ignore if one is already set), or adding
+    // TLS client certificate credentials.
+    UpdateTransportConfig(c *transport.Config)
+
+    // Login() dropped, it was never used.
+}
+```
+
+This would let auth transports supply TLS credentials, as well as instrument
+transports with in-memory rotation code like the utilities implemented by
+[`k8s.io/client-go/util/certificate`](https://godoc.org/k8s.io/client-go/util/certificate).
+
+If the exec command specifies `kind: tls-certificate`, the command is expected to
+output a unencrypted PEM encoded private key, followed by the PEM encoded client
+certificate chain.
+
+```
+-----BEGIN RSA PRIVATE KEY-----
+MIIEowIBAAKCAQEA6IVXGPX5yP2Q6TAlQXIQsavzSqZ973iZvpQBGTI6M98gTSVm
+# ...
+TUwpZmylTKEJ9zLt2PADglyDrQ2D+1WNzh966Oo9c+kZt4WJM0aF
+-----END RSA PRIVATE KEY-----
+-----BEGIN CERTIFICATE-----
+MIIDCzCCAfOgAwIBAgIJAKK9m2Cfg5uhMA0GCSqGSIb3DQEBCwUAMBsxGTAXBgNV
+# ...
+xar2YeJ6mCCzSAPM69DP
+-----END CERTIFICATE-----
+-----BEGIN CERTIFICATE-----
+MIIDCzCCAfOgAwIBAgIJAPqJyUfmRxGLMA0GCSqGSIb3DQEBCwUAMBsxGTAXBgNV
+# ...
+zgmWLORg+ls1H1oaJiNW
+-----END CERTIFICATE-----
+```
+
+The `AuthProvider` then adds those credentials to the `transport.Config`.
+
+## Login
+
+Historically, `AuthProviders` have had a `Login()` method with the hope that it
+could trigger bootstrapping into the cluster. While no providers implement this
+method, the Azure `AuthProvider` can already prompt an [interactive auth flow](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.8.5/plugin/pkg/client/auth/azure/azure.go#L343).
+This suggests that an exec'd tool should be able to trigger its own custom logins,
+either by opening a browser, or performing a text based prompt.
+
+We should take care that interactive stderr and stdin are correctly
+inherited by the sub-process to enable this kind of interaction.
```

Comment:
How would you distinguish between the cases when the authenticator process hangs or waits for the user input (e.g. interactive flow)?  I guess a sort of execution timeout of the subprocess is required. On the other hand, kubectl has more control over this in case of a plugin.

Reply:
I'd like to start with the subprocess implementing this logic and present appropriate feedback.

If this is interactive, the user will likely CTRL+c the execution anyway.

Reply:
Made this clear.